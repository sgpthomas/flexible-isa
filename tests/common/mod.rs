use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
