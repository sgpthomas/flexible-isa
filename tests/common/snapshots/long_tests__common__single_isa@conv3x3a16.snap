---
source: tests/common/mod.rs
description: generate isa for conv3x3a16
---
[
    (
        69,
        1,
        "(+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#17 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#18 (cast<int32> 1))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#19 (cast<unknownx128> (call[ramp] (cast<unknown> ?x2) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#20 (cast<int32> 0))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#21 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#22 (cast<int32> 2))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#23 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#24 (cast<int32> 3))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#25 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#26 (cast<int32> 4))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#27 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#28 (cast<int32> 5))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#29 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#30 (cast<int32> 6))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#31 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#32 (cast<int32> 8))))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#33 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x11))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#34 (cast<int32> 7))))))))))))))))))))))",
    ),
    (
        85,
        1,
        "(+ (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        87,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 6))",
    ),
    (
        88,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 8))",
    ),
    (
        89,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 13))",
    ),
    (
        92,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (- (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?#4))) (cast<int32> 4))) (cast<int32> ?x0))) (cast<int32> ?#5))) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> 1))",
    ),
    (
        93,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?#4))) (cast<int32> 4))) (cast<int32> ?x0))) (cast<int32> ?#5))) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        94,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        96,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        102,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        104,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 5))",
    ),
    (
        109,
        1,
        "(+ (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<unknown> (* (cast<int32> (+ (cast<int32> (- (cast<int32> ?#1) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 5))))))) (cast<int32> 1))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#4) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))))) (cast<int32> 128))",
    ),
    (
        111,
        1,
        "(+ (cast<unknown> (* (cast<unknown> (+ (cast<unknown> ?#1) (cast<unknown> (call[max] (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> (call[min] (cast<unknown> (+ (cast<unknown> ?#2) (cast<int32> 254))) (cast<int32> ?#3))) (cast<int32> (neg (cast<int32> 127))))) (cast<int32> 128))) (cast<int32> 0))))) (cast<int32> 128))) ?x0)",
    ),
    (
        130,
        1,
        "(+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<unknown> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> 1))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#2) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))))",
    ),
    (
        133,
        1,
        "(+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        138,
        1,
        "(+ (cast<unknown> ?x0) (cast<int32> 127))",
    ),
    (
        139,
        1,
        "(+ (cast<unknown> ?x0) (cast<int32> 128))",
    ),
    (
        158,
        1,
        "(+ ?x0 (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#2))))))",
    ),
    (
        166,
        1,
        "(- ?x0 ?x1)",
    ),
    (
        168,
        1,
        "(* (cast<int32> (+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#3))",
    ),
    (
        171,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#1))",
    ),
    (
        172,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        175,
        1,
        "(* (cast<int32> ?x0) (cast<int32> 4))",
    ),
    (
        182,
        1,
        "(* (cast<unknown> ?x0) (cast<int32> 4))",
    ),
    (
        184,
        1,
        "(* (cast<unknown> ?x0) (cast<int32> ?x1))",
    ),
    (
        199,
        1,
        "(<= ?x0 (cast<int32> ?x1))",
    ),
    (
        201,
        1,
        "(== ?#0 ?#1)",
    ),
    (
        202,
        1,
        "(&& ?#0 ?#1)",
    ),
    (
        204,
        1,
        "(get ?x0 (cast<unknownx128> (call[ramp] (cast<unknown> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        205,
        1,
        "(get ?x0 ?x1)",
    ),
    (
        221,
        1,
        "(get ?#3 (cast<unknown> (+ (cast<unknownx128> (call[max] (cast<unknownx128> (call[min] (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#4) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x2) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#5))))))))",
    ),
    (
        86,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 5))",
    ),
    (
        100,
        2,
        "(+ (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (neg (cast<int32> ?x2))))",
    ),
    (
        105,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 127))",
    ),
    (
        106,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        136,
        2,
        "(+ (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        137,
        2,
        "(+ (cast<unknown> ?x0) (cast<int32> 1))",
    ),
    (
        164,
        2,
        "(- (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))))",
    ),
    (
        169,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))",
    ),
    (
        183,
        2,
        "(* (cast<unknown> ?x0) (cast<int32> 128))",
    ),
    (
        190,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        193,
        2,
        "(/ ?x0 (cast<int32> 4))",
    ),
    (
        194,
        2,
        "(/ ?x0 (cast<int32> 128))",
    ),
    (
        207,
        2,
        "(get ?#1 (cast<unknown> (+ (cast<unknownx128> (call[max] (cast<unknownx128> (call[min] (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?#2) (cast<int32> 128))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#3) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?#4))))))))",
    ),
    (
        76,
        3,
        "(+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#18 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#19 (cast<int32> 1))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#20 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#21 (cast<int32> 0))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#22 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#23 (cast<int32> 2))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#24 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#25 (cast<int32> 3))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#26 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#27 (cast<int32> 4))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#28 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#29 (cast<int32> 5))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#30 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#31 (cast<int32> 6))))))) (cast<int16x128> (+ (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#32 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#33 (cast<int32> 8))))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#34 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x11))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> (get ?#35 (cast<int32> 7))))))))))))))))))))))",
    ),
    (
        98,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#2) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 6))))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        99,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?x0))) (cast<int32> 4))) (cast<int32> 6))))) (cast<int32> ?x1))",
    ),
    (
        103,
        3,
        "(+ (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        112,
        3,
        "(+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> ?x0))",
    ),
    (
        141,
        3,
        "(+ (cast<unknown> ?x0) (cast<int32> ?x1))",
    ),
    (
        163,
        3,
        "(- (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?#2))) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 5))))))",
    ),
    (
        177,
        3,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        198,
        3,
        "(<= (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 1))) (cast<int32> ?x3))",
    ),
    (
        108,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        128,
        4,
        "(+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        142,
        4,
        "(+ (cast<unknown> ?x0) (cast<unknown> ?x1))",
    ),
    (
        174,
        5,
        "(* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))",
    ),
    (
        176,
        6,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        0,
        7,
        "(neg (cast<int32> ?x0))",
    ),
    (
        189,
        8,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        113,
        9,
        "(+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> ?x0))) (cast<unknown> ?#2))",
    ),
    (
        191,
        9,
        "(/ (cast<unknown> (+ (cast<unknown> ?x0) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
]
